metricbeat/module/kafka/partition/partition.go

```diff
@@ -19,75 +25,292 @@ func init() {
+
+	// current broker is bootstrap only. Get metadata to find id:
+	meta, err := queryMetadataWithRetry(b, m.cfg, m.topics)
+	if err != nil {
+		closeBroker(b)
+		return nil, err
+	}
+
+	addr := b.Addr()
+	for _, other := range meta.Brokers {
+		if other.Addr() == addr {
+			m.id = other.ID()
+			break
+		}
+	}
+
+	if m.id == noID {
+		closeBroker(b)
+		err = fmt.Errorf("No advertised broker with address %v found", addr)
+		return nil, err
+	}
+
+	return b, nil
 }
 
 // Fetch partition stats list from kafka
 func (m *MetricSet) Fetch() ([]common.MapStr, error) {
-	if m.client == nil {
-		config := sarama.NewConfig()
-		config.Net.DialTimeout = m.Module().Config().Timeout
-		config.Net.ReadTimeout = m.Module().Config().Timeout
-		config.ClientID = "metricbeat"
-
-		client, err := sarama.NewClient([]string{m.Host()}, config)
-		if err != nil {
-			return nil, err
-		}
-		m.client = client
+	b, err := m.connect()
+	if err != nil {
+		return nil, err
 	}
 
-	topics, err := m.client.Topics()
+	defer closeBroker(b)
+	response, err := queryMetadataWithRetry(b, m.cfg, m.topics)
 	if err != nil {
 		return nil, err
 	}
 
 	events := []common.MapStr{}
-	for _, topic := range topics {
-		partitions, err := m.client.Partitions(topic)
-		if err != nil {
-			logp.Err("Fetch partition info for topic %s: %s", topic, err)
+	evtBroker := common.MapStr{
+		"id":      m.id,
+		"address": b.Addr(),
+	}
+
+	for _, topic := range response.Topics {
+		evtTopic := common.MapStr{
+			"name": topic.Name,
+			"error": common.MapStr{
+				"code": topic.Err,
+			},
 		}
 
-		for _, partition := range partitions {
-			newestOffset, err := m.client.GetOffset(topic, partition, sarama.OffsetNewest)
-			if err != nil {
-				logp.Err("Fetching newest offset information for partition %s in topic %s: %s", partition, topic, err)
+		for _, partition := range topic.Partitions {
+			// partition offsets can be queried from leader only
+			if m.id != partition.Leader {
+				continue
 			}
 
-			oldestOffset, err := m.client.GetOffset(topic, partition, sarama.OffsetOldest)
-			if err != nil {
-				logp.Err("Fetching oldest offset information for partition %s in topic %s: %s", partition, topic, err)
-			}
+			// collect offsets for all replicas
+			for _, id := range partition.Replicas {
+
+				// Get oldest and newest available offsets
+				offOldest, offNewest, offOK, err := queryOffsetRange(b, id, topic.Name, partition.ID)
+
+				if !offOK {
+					if err == nil {
+						err = errFailQueryOffset
+					}
 
-			broker, err := m.client.Leader(topic, partition)
-			if err != nil {
-				logp.Err("Fetching brocker for partition %s in topic %s: %s", partition, topic, err)
+					logp.Err("Failed to query kafka partition (%v:%v) offsets: %v",
+						topic.Name, partition.ID, err)
+					continue
+				}
+
+				// create event
+				event := common.MapStr{
+					"topic":  evtTopic,
+					"broker": evtBroker,
+					"partition": common.MapStr{
+						"id": partition.ID,
+						"error": common.MapStr{
```

Comment:
values `error.code: 0` is a valid error code in kafka. That's why I did keep it in.